pipeline-scripts/commonlib.groovy

```diff
@@ -236,17 +244,23 @@ def shell(arg) {
     def script = arg.script  // keep a copy of original script
     def truncScript = (script.size() <= 75) ? script : script[0..35] + "..." + script[-36..-1]
 
-    // ensure a clean space to save output files
-    if (!shellIndex++) { sh("rm -rf shell") }
-
-    def filename = "shell/sh.${shellIndex}." + truncScript.replaceAll( /[^\w-.]+/ , "_").take(80)
-    sh("mkdir -p ${filename}")
-    filename += "/sh.${shellIndex}"
+    // put shell result files in a subdir specific to this shell call
+    def threadShellIndex = getNextShellCounter()
+    def filename = "shell/sh.${threadShellIndex}." + truncScript.replaceAll( /[^\w-.]+/ , "_").take(80)
+    filename += "/sh.${threadShellIndex}"
+
+    // concurrency-safe creation of output dir and removal of previous
+    def shellDir = "shell.${currentBuild.number}"
+    sh """#!/bin/bash +x
+        mkdir -p ${shellDir}
+        ln -sfn ${shellDir} shell
+        mkdir -p ${filename}
+        rm -rf shell.foo `ls -d shell.* | sed -e '/^${shellDir}\$/d' `
```

Comment:
`rm -rf` in my experience is a safe rc=0 operation when no arguments are supplied.

Reply:
hahah, i can't believe i didn't try it to see!